src/job_runner/job_runner.cc

```diff
@@ -0,0 +1,192 @@
+}
+
+std::size_t JobRunner::DefaultSchedulerHint() {
+    std::uniform_int_distribution<std::size_t> random_worker_selector(0, config_.thread_num_);
+
+    return kCurrentThreadJobRunner.load() == reinterpret_cast<std::uintptr_t>(this)
+        ? kCurrentThreadWorkerIndex.load()
+        : random_worker_selector(random_engine);
+}
+
+JobRunner::Worker::Worker(JobRunner* runner, std::size_t idx)
+    : runner_(runner)
+    , index_(idx)
+    , thread_(std::bind(&Worker::WorkerLoop, this)) {
+}
+
+JobRunner::Worker::~Worker() {
+    DCHECK(!thread_.joinable());
+    job_queue_.consume_all([](job_t* const job_ptr) { delete job_ptr; });
+}
+
+std::unique_ptr<JobRunner::job_t> JobRunner::Worker::TryGetOneJob() {
+    std::unique_ptr<job_t> job{nullptr};
+    job_queue_.consume_one([&job](job_t* const job_ptr) { job.reset(job_ptr); });
+    return job;
+}
+
+bool JobRunner::Worker::TryProcessOne() {
+    if (auto job = TryGetOneJob()) {
+        (*job)();
+        return true;
+    }
+    return false;
+}
+
+void JobRunner::Worker::WorkerLoop() {
+    cr_timestamp_nsec_t last_active_timestamp = GetSystemTimestampNsec();
+
+    const long long active_time_nsec =
+        std::chrono::duration_cast<std::chrono::nanoseconds>(runner_->config_.active_time_).count();
+
+    kCurrentThreadJobRunner.store(reinterpret_cast<std::uintptr_t>(runner_));
+    kCurrentThreadWorkerIndex.store(index_);
+
+    runner_->active_workers_num_.fetch_add(1);
+    while (!shutdown_flag_.load()) {
+        if (TryProcessOne() || runner_->Steal()) {
+            last_active_timestamp = GetSystemTimestampNsec();
+            continue;
+        }
+        if (index_ < runner_->config_.always_active_thread_num_) {
+            continue;
+        }
+        if (GetSystemTimestampNsec() < last_active_timestamp + active_time_nsec) {
+            continue;
+        }
+        runner_->active_workers_num_.fetch_sub(1);
+        DLOG(INFO) << __func__ << ": Worker " << index_ << " is inactive.";
+        std::unique_lock<std::mutex> lock(inactive_cv_mutex_);
+        constexpr auto               kWorkerIdleTime = std::chrono::seconds(1);
+        inactive_cv_.wait_for(lock, kWorkerIdleTime);
+        DLOG(INFO) << __func__ << ": Worker " << index_ << " is active.";
+        runner_->active_workers_num_.fetch_add(1);
+    }
+    runner_->active_workers_num_.fetch_sub(1);
+    stopped_flag_.store(true);
+}
+
+void JobRunner::Worker::Stop() {
+    bool expected_shutdown_flag = false;
+    if (!shutdown_flag_.compare_exchange_strong(expected_shutdown_flag, true)) {
+        return;
+    }
+    inactive_cv_.notify_all();
+    DLOG(INFO) << __func__ << ": Worker " << index_ << " is stopping.";
+}
+
+void JobRunner::Worker::Join() {
+    while (!stopped_flag_.load()) {
+        inactive_cv_.notify_all();
+    }
+    DLOG(INFO) << __func__ << ": Worker " << index_ << " is stopped.";
+    thread_.join();
+    std::atomic_thread_fence(std::memory_order::seq_cst);
```

Comment:
@xkszltl It **almost** works because it synced with the last `stopped_flag_` setting in the worker threads. However if you run ubsan tests many times you may still find some issues like this:

```
INFO: From Testing //tests:job_runner_test:
==================== Test output for //tests:job_runner_test:
Running main() from gmock_main.cc
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from JobRunnerTest
[ RUN      ] JobRunnerTest.Basic
UndefinedBehaviorSanitizer:DEADLYSIGNAL
==12==ERROR: UndefinedBehaviorSanitizer: SEGV on unknown address 0xffffffffffffff82 (pc 0x7f7c60ee3021 bp 0x7fffea609890 sp 0x7fffea609880 T12)
==12==The signal is caused by a READ memory access.
    #0 0x7f7c60ee3021 in cris::core::JobRunner::Worker::Join() /proc/self/cwd/src/job_runner/job_runner.cc
    #1 0x7f7c60ee2edc in cris::core::JobRunner::~JobRunner() /proc/self/cwd/src/job_runner/job_runner.cc:36:17
    #2 0x22d1cf in cris::core::JobRunnerTest_Basic_Test::TestBody() /proc/self/cwd/tests/job_runner_test.cc:86:1
    #3 0x7f7c60d4735e in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (/home/
chenhao/.cache/bazel/_bazel_chenhao/dd7b92a42201a8d2f5ee22f4696125c0/execroot/cris-core/bazel-out/k8-opt/bin/tests/../_solib_k8/libexternal_Sgtest_Slibgtest.so+0x13335e)
    #4 0x7f7c60d47029 in testing::Test::Run() (/home/chenhao/.cache/bazel/_bazel_chenhao/dd7b92a42201a8d2f5ee22f4696125c0/execroot/cris-core/bazel-out/k8-opt/bin/tests/.
./_solib_k8/libexternal_Sgtest_Slibgtest.so+0x133029)                               
    #5 0x7f7c60d48f09 in testing::TestInfo::Run() (/home/chenhao/.cache/bazel/_bazel_chenhao/dd7b92a42201a8d2f5ee22f4696125c0/execroot/cris-core/bazel-out/k8-opt/bin/tes
ts/../_solib_k8/libexternal_Sgtest_Slibgtest.so+0x134f09)                           
    #6 0x7f7c60d4b47e in testing::TestCase::Run() (/home/chenhao/.cache/bazel/_bazel_chenhao/dd7b92a42201a8d2f5ee22f4696125c0/execroot/cris-core/bazel-out/k8-opt/bin/tes
ts/../_solib_k8/libexternal_Sgtest_Slibgtest.so+0x13747e)                           
    #7 0x7f7c60d5f567 in testing::internal::UnitTestImpl::RunAllTests() (/home/chenhao/.cache/bazel/_bazel_chenhao/dd7b92a42201a8d2f5ee22f4696125c0/execroot/cris-core/ba
zel-out/k8-opt/bin/tests/../_solib_k8/libexternal_Sgtest_Slibgtest.so+0x14b567)     
    #8 0x7f7c60d5e389 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testi
ng::internal::UnitTestImpl::*)(), char const*) (/home/chenhao/.cache/bazel/_bazel_chenhao/dd7b92a42201a8d2f5ee22f4696125c0/execroot/cris-core/bazel-out/k8-opt/bin/tests/
../_solib_k8/libexternal_Sgtest_Slibgtest.so+0x14a389)                              
    #9 0x7f7c60d5e02f in testing::UnitTest::Run() (/home/chenhao/.cache/bazel/_bazel_chenhao/dd7b92a42201a8d2f5ee22f4696125c0/execroot/cris-core/bazel-out/k8-opt/bin/tes
ts/../_solib_k8/libexternal_Sgtest_Slibgtest.so+0x14a02f)                           
    #10 0x7f7c60dcecd5 in main (/home/chenhao/.cache/bazel/_bazel_chenhao/dd7b92a42201a8d2f5ee22f4696125c0/execroot/cris-core/bazel-out/k8-opt/bin/tests/../_solib_k8/lib
external_Sgtest_Slibgtest_Umain.so+0x1cd5)
    #11 0x7f7c6067b082 in __libc_start_main /build/glibc-SzIz7B/glibc-2.31/csu/../csu/libc-start.c:308:16
    #12 0x21687d in _start (/home/chenhao/.cache/bazel/_bazel_chenhao/dd7b92a42201a8d2f5ee22f4696125c0/execroot/cris-core/bazel-out/k8-opt/bin/tests/job_runner_test+0x21
687d)                                     

UndefinedBehaviorSanitizer can not provide additional info.
SUMMARY: UndefinedBehaviorSanitizer: SEGV /proc/self/cwd/src/job_runner/job_runner.cc in cris::core::JobRunner::Worker::Join()
==12==ABORTING                            
================================================================================
```

Reply:
1. Maybe joinable isn't reliable? Your Join() is supposed to be noop.
2. addr2line?

Reply:
1. joinable is not threadsafe, but we are not joining from multiple threads but only the main thread.
2. the end of each line of stack trace gives you addr2line info.